### ISSUE TITLE
Vmapped QR on GPU is not parallelized

### ISSUE THREAD

**Issue by AdrienCorenflos** [google/jax#8542]
Hi,

I just noticed that QR decomposition batching rule does not work on GPU: in fact the runtime increases linearly with the size of the batch. This is illustrated below. I have written a small reproduction in colab that you can find [here](https://colab.research.google.com/drive/1LnfEHAs_B3h7D8JAjd_EhAZTNHfjQ7fO#scrollTo=mXAtWZPQFAON).

Let me know if you need more info!
Adrien

![QR](https://user-images.githubusercontent.com/19948263/141806336-d8aeed28-bd9b-447e-949e-7ad420465f13.png)
![cholesky](https://user-images.githubusercontent.com/19948263/141806346-047cc8e4-8f1b-4ac9-ba84-a1737ce8bde4.png)

**Comment by hawkinsp** [comment]
The issue is that Nvidia do not provide a batched QR decomposition in Cusolver, but they *do* provide a batched Cholesky decomposition. So we implement batched QR by looping on GPU (I think we implement batched QR by looping on all platforms).

It's perfectly possible to implement a batched QR decomposition, but someone (either us or Nvidia) would have to do it.

It's possible that writing out a QR decomposition in Python and `jit` compiling it might work well for such a small matrix.

**Comment by hawkinsp** [comment]
By the way, if you wanted to try that last part, there's a pure NumPy implementation in the comments of (part of ) the TPU implementation:

https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/qr_expander.cc;l=177

You don't need blocking for a 5x5 matrix, so you need just the inner QR kernel and the code to compute a Householder reflector.

**Comment by AdrienCorenflos** [comment]
To give some context, I'm working on a paper that does parallel-in-time sqrt (non linear) kalman filtering and smoothing, so we need to do triangularization operations (tria) that are easily implemented in parallel. Because I'm using sigma-point integration methods (cubature, unscented, etc), the matrix largest dimension will be around 10 to 15 I think. Do you think this Python impl still would work out of the box?

**Comment by hawkinsp** [comment]
I think it would work (produce correct outputs). The open questions are:
* how long does it take to compile (for 15x15, very reasonable I expect, even with fully unrolled loops)?
* how fast is it?

For the latter question: I have no idea without trying it. If you wanted to try it out and report back, that'd be informative. We could make the JAX QR decomposition implementation do this internally if it turns out to be a big win for some shapes.

**Comment by hawkinsp** [comment]
I think both are useful. So likely what we would want to do is choose an implementation based on the matrix size and the batch size, using this kind of implementation only where it is faster.

**Comment by AdrienCorenflos** [comment]
I ran a comparison on [colab](https://colab.research.google.com/drive/1wfrNg9IhqoalTiRca78o1mMP2PXrt5f1?usp=sharing), the charts below are what I get. The title font is a bit small (my colab session expired), so for the sake of readibility TOP row in a chart is the jitted `jnp.linalg.qr`, and bottom row in a chart is the function [above](https://github.com/google/jax/issues/8542#issuecomment-969113921) (with householder) jitted too.

First, **compilation time** as a function of the matrix shape:

On CPU            |  On GPU
:-------------------------:|:-------------------------:
  ![compile_cpu](https://user-images.githubusercontent.com/19948263/141985440-cacc1890-4aa6-4f98-8737-706dd23c406f.png) | ![compile_gpu](https://user-images.githubusercontent.com/19948263/141985475-e3471920-b044-473d-b751-08bb2146e4ec.png)


Second, **run time** as a function of the matrix shape:

On CPU            |  On GPU
:-------------------------:|:-------------------------:
![runtime_cpu](https://user-images.githubusercontent.com/19948263/141985572-3ed815d6-444c-44a7-b3e5-de0d0cb40a53.png) | ![runtime_gpu](https://user-images.githubusercontent.com/19948263/141985588-ff4e8c73-5d90-47c9-b80d-ad05cd5ed28f.png)

**Comment by AdrienCorenflos** [comment]
I guess the conclusion is "if one of the dimensions of the matrix is less than 10, it does not really matter which method you use".
I have however not checked memory consumption which may also have an impact.

**Comment by AdrienCorenflos** [comment]
I noticed that the householder above is unstable for degenerate matrices (like np.zeros). This version is more general

```python
def _householder(a, eps=1e-6):
    alpha = a[0]
    s = jnp.sum(a[1:] ** 2)
    cond = s < eps

    def if_not_cond(v):
        t = (alpha ** 2 + s) ** 0.5
        v0 = jax.lax.cond(alpha <= 0, lambda _: alpha - t, lambda _: -s / (alpha + t), None)
        tau = 2 * v0 ** 2 / (s + v0 ** 2)
        v = v / v0
        v = v.at[0].set(1.)
        return v, tau

    return jax.lax.cond(cond, lambda v: (v, 0.), if_not_cond, a)
```

**Comment by mj023** [comment]
I encountered a similar problem. The reason this is implemented this way seems to be that calculating R and a representation of Q as Householder Vectors is batched in the Cusolver Library, but actually calculating Q, which is called the Householder-Product in JAX, is not. For large batches of small matrices there is a lot of communication between the CPU and GPU happening, slowing down the computation. If you just need R, JAX will optimize away the Housholder Product. The gradient however is calculated using Q, so it can't be optimized away and Q will be calculated even if you do not use it in your main computation. 

The workaround that I found was using jax.numpy.linalg.qr with mode='raw' to calculate R and then implementing the Housholder Product in JAX. If you want to use the gradient, you would have to implement a custom jvp for this function (Modifying the current JVP for QR a bit works fine).

```python
def qr_gpu(a: jax.Array):

    r, tau = jnp.linalg.qr(a, mode="raw")
    q = _householder(r.mT, tau)
    return q, jnp.triu(r.mT[: tau.shape[0]])

def _householder(r: jax.Array, tau: jax.Array):
    m = r.shape[0]
    n = tau.shape[0]
    r = jnp.tril(jnp.fill_diagonal(r, 1, inplace=False))
    # Calculate Householder Vector which is saved in the lower triangle of R
    v1 = jnp.expand_dims(r[:, 0], 1)
    h = jnp.eye(m) - tau[0] * (v1 @ jnp.conj(jnp.transpose(v1)))
    # Multiply all Householder Vectors Q = H(1)*H(2)...*H(n)
    for i in range(1, n):
        vi = jnp.expand_dims(r[:, i], 1)
        h = h - tau[i] * (h @ vi) @ jnp.conj(jnp.transpose(vi))
    return h[:, :n]
```
I do not know if this is exactly how they calculate Q in Cusolver, but for my purpose, a Square Root Kalman Filter, it worked alright and was a lot faster.

**Comment by yantaow** [comment]
> I encountered a similar problem. The reason this is implemented this way seems to be that calculating R and a representation of Q as Householder Vectors is batched in the Cusolver Library, but actually calculating Q, which is called the Householder-Product in JAX, is not. For large batches of small matrices there is a lot of communication between the CPU and GPU happening, slowing down the computation. If you just need R, JAX will optimize away the Housholder Product. The gradient however is calculated using Q, so it can't be optimized away and Q will be calculated even if you do not use it in your main computation.
> 
> The workaround that I found was using jax.numpy.linalg.qr with mode='raw' to calculate R and then implementing the Housholder Product in JAX. If you want to use the gradient, you would have to implement a custom jvp for this function (Modifying the current JVP for QR a bit works fine).
> 
> def qr_gpu(a: jax.Array):
> 
>     r, tau = jnp.linalg.qr(a, mode="raw")
>     q = _householder(r.mT, tau)
>     return q, jnp.triu(r.mT[: tau.shape[0]])
> 
> def _householder(r: jax.Array, tau: jax.Array):
>     m = r.shape[0]
>     n = tau.shape[0]
>     # Calculate Householder Vector which is saved in the lower triangle of R
>     v1 = jnp.expand_dims(r[:, 0], 1)
>     v1 = v1.at[0:0].set(0)
>     v1 = v1.at[0].set(1)
>     h = jnp.eye(m) - tau[0] * (v1 @ jnp.transpose(v1))
>     # Multiply all Householder Vectors Q = H(1)*H(2)...*H(n)
>     for i in range(1, n):
>         vi = jnp.expand_dims(r[:, i], 1)
>         vi = vi.at[0:i].set(0)
>         vi = vi.at[i].set(1)
>         h = h @ (jnp.eye(m) - tau[i] * (vi @ jnp.transpose(vi)))
>     return h[:, :n]
> I do not know if this is exactly how they calculate Q in Cusolver, but for my purpose, a Square Root Kalman Filter, it worked alright and was a lot faster.

oh my friend, you are my hero. This works very well!!

**Comment by fliingelephant** [comment]
> I encountered a similar problem. The reason this is implemented this way seems to be that calculating R and a representation of Q as Householder Vectors is batched in the Cusolver Library, but actually calculating Q, which is called the Householder-Product in JAX, is not. For large batches of small matrices there is a lot of communication between the CPU and GPU happening, slowing down the computation. If you just need R, JAX will optimize away the Housholder Product. The gradient however is calculated using Q, so it can't be optimized away and Q will be calculated even if you do not use it in your main computation.
> 
> The workaround that I found was using jax.numpy.linalg.qr with mode='raw' to calculate R and then implementing the Housholder Product in JAX. If you want to use the gradient, you would have to implement a custom jvp for this function (Modifying the current JVP for QR a bit works fine).
> 
> def qr_gpu(a: jax.Array):
> 
>     r, tau = jnp.linalg.qr(a, mode="raw")
>     q = _householder(r.mT, tau)
>     return q, jnp.triu(r.mT[: tau.shape[0]])
> 
> def _householder(r: jax.Array, tau: jax.Array):
>     m = r.shape[0]
>     n = tau.shape[0]
>     # Calculate Householder Vector which is saved in the lower triangle of R
>     v1 = jnp.expand_dims(r[:, 0], 1)
>     v1 = v1.at[0:0].set(0)
>     v1 = v1.at[0].set(1)
>     h = jnp.eye(m) - tau[0] * (v1 @ jnp.transpose(v1))
>     # Multiply all Householder Vectors Q = H(1)*H(2)...*H(n)
>     for i in range(1, n):
>         vi = jnp.expand_dims(r[:, i], 1)
>         vi = vi.at[0:i].set(0)
>         vi = vi.at[i].set(1)
>         h = h @ (jnp.eye(m) - tau[i] * (vi @ jnp.transpose(vi)))
>     return h[:, :n]
> I do not know if this is exactly how they calculate Q in Cusolver, but for my purpose, a Square Root Kalman Filter, it worked alright and was a lot faster.

Thanks for sharing this implementation. One small note: for complex-valued inputs, the Householder formula should use the conjugate transpose, not a plain transpose.

Following your implementation, I implemented a batch-polymorphic Householder QR using the compact WY representation as well. See https://github.com/jax-ml/jax/discussions/31470#discussioncomment-15694094. In my tests, the compact WY way seems to be faster.

**Comment by mj023** [comment]
Thank you for pointing this out, I did not even think of complex numbers. Your implementation helped me improve the runtime a bit further and I also did a little benchmarking. I am facing very small matrices < (10x10) and for these matrices the runtime of my implementation seems to be better. For larger matrices though, the WY - variant and the for-loop are similar, even though I ran out of memeory faster with the WY- variant. The biggest drawback of the for-loop-variant is that compile time scales with the dimensions of the matrices, which can get out of hand quickly.

|  Matrix size  |    WY   | For Loop |
|:-------------:|:-------:|:--------:|
| 10000x(10x10) | 15.68ms |  3.72ms  |
| 10000x(64x64) |  451ms  |   378ms  |
| 10000x(96x96) |   OOM   |   1.1s   |

Which implementation fits better, seems to really depend on the kind of problem one is facing.
